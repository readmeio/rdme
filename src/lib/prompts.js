const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
