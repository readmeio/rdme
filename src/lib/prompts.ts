import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
