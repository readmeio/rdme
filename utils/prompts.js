exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
