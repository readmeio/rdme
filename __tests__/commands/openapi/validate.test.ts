import fs from 'node:fs';

import prompts from 'prompts';
import { describe, beforeAll, beforeEach, afterEach, it, expect, vi } from 'vitest';

import Command from '../../../src/commands/openapi/validate.js';
import { after, before } from '../../helpers/get-gha-setup.js';
import { runCommand, runCommandWithHooks, type OclifOutput } from '../../helpers/oclif.js';

describe('rdme openapi validate', () => {
  let run: (args?: string[]) => OclifOutput;
  let testWorkingDir: string;

  beforeAll(() => {
    run = runCommand(Command);
  });

  beforeEach(() => {
    testWorkingDir = process.cwd();
  });

  afterEach(() => {
    process.chdir(testWorkingDir);
  });

  it.each([
    ['Swagger 2.0', 'json', '2.0'],
    ['Swagger 2.0', 'yaml', '2.0'],
    ['OpenAPI 3.0', 'json', '3.0'],
    ['OpenAPI 3.0', 'yaml', '3.0'],
    ['OpenAPI 3.1', 'json', '3.1'],
    ['OpenAPI 3.1', 'yaml', '3.1'],
  ])('should support validating a %s definition (format: %s)', async (_, format, specVersion) => {
    expect(
      (await run([require.resolve(`@readme/oas-examples/${specVersion}/${format}/petstore.${format}`)])).result,
    ).toContain(`petstore.${format} is a valid ${specVersion === '2.0' ? 'Swagger' : 'OpenAPI'} API definition!`);
  });

  it('should discover and upload an API definition if none is provided', async () => {
    const result = await run(['--workingDirectory', './__tests__/__fixtures__/relative-ref-oas']);
    expect(result).toMatchSnapshot();
  });

  it('should use specified working directory', async () => {
    const result = await run(['petstore.json', '--workingDirectory', './__tests__/__fixtures__/relative-ref-oas']);
    expect(result).toMatchSnapshot();
  });

  it('should adhere to .gitignore in subdirectories', async () => {
    fs.copyFileSync(
      require.resolve('@readme/oas-examples/3.0/json/petstore-simple.json'),
      './__tests__/__fixtures__/nested-gitignored-oas/nest/petstore-ignored.json',
    );

    const result = await run(['--workingDirectory', './__tests__/__fixtures__/nested-gitignored-oas']);
    expect(result).toMatchSnapshot();
  });

  describe('error handling', () => {
    it('should throw an error if invalid JSON is supplied', () => {
      return expect(run(['./__tests__/__fixtures__/invalid-json/yikes.json'])).resolves.toMatchSnapshot();
    });

    it('should throw an error if an invalid OpenAPI 3.0 definition is supplied', () => {
      return expect(run(['./__tests__/__fixtures__/invalid-oas.json'])).resolves.toMatchSnapshot();
    });

    it('should throw an error if an invalid OpenAPI 3.1 definition is supplied', () => {
      return expect(run(['./__tests__/__fixtures__/invalid-oas-3.1.json'])).resolves.toMatchSnapshot();
    });

    it('should throw an error if an invalid Swagger definition is supplied', () => {
      return expect(run(['./__tests__/__fixtures__/invalid-swagger.json'])).resolves.toMatchSnapshot();
    });

    it('should throw an error if an invalid API definition has many errors', () => {
      return expect(run(['./__tests__/__fixtures__/very-invalid-oas.json'])).resolves.toMatchSnapshot();
    });
  });

  describe('CI tests', () => {
    beforeEach(() => {
      vi.stubEnv('TEST_RDME_CI', 'true');
    });

    afterEach(() => {
      vi.unstubAllEnvs();
    });

    it('should successfully validate prompt and not run GHA onboarding', async () => {
      vi.stubEnv('TEST_RDME_CREATEGHA', 'true');
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';

      await expect(run([spec])).resolves.toMatchSnapshot();
    });

    it('should fail if user attempts to pass `--github` flag in CI environment', async () => {
      return expect(
        (
          await runCommandWithHooks([
            'openapi validate',
            '__tests__/__fixtures__/petstore-simple-weird-version.json',
            '--github',
          ])
        ).error,
      ).toMatchSnapshot();
    });
  });

  describe('GHA onboarding E2E tests', () => {
    let yamlOutput;

    beforeEach(() => {
      before((fileName, data) => {
        yamlOutput = data;
      });
    });

    afterEach(() => {
      after();
    });

    it('should create GHA workflow if user passes in spec via prompts', async () => {
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      const fileName = 'validate-test-file';
      prompts.inject([spec, true, 'validate-test-branch', fileName]);

      await expect(run()).resolves.toMatchSnapshot();
      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should create GHA workflow if user passes in spec via opt', async () => {
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      const fileName = 'validate-test-opt-spec-file';
      prompts.inject([true, 'validate-test-opt-spec-branch', fileName]);

      await expect(run([spec])).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should create GHA workflow if user passes in spec via opt (including workingDirectory)', async () => {
      const spec = 'petstore.json';
      const fileName = 'validate-test-opt-spec-workdir-file';
      prompts.inject([true, 'validate-test-opt-spec-github-branch', fileName]);

      await expect(
        run([spec, '--workingDirectory', './__tests__/__fixtures__/relative-ref-oas']),
      ).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should create GHA workflow if user passes in spec via opt (github flag enabled)', async () => {
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      const fileName = 'validate-test-opt-spec-github-file';
      prompts.inject(['validate-test-opt-spec-github-branch', fileName]);

      await expect(run([spec, '--github'])).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should reject if user says no to creating GHA workflow', () => {
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      prompts.inject([spec, false]);
      return expect(run()).resolves.toMatchSnapshot();
    });
  });
});
