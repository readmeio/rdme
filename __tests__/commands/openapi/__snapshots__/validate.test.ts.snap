// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rdme openapi validate > CI tests > should fail if user attempts to pass \`--github\` flag in CI environment 1`] = `
[Error: Parsing --github 
	The \`--github\` flag is only for usage in non-CI environments.
See more help with --help]
`;

exports[`rdme openapi validate > CI tests > should successfully validate prompt and not run GHA onboarding 1`] = `
{
  "result": "__tests__/__fixtures__/petstore-simple-weird-version.json is a valid OpenAPI API definition!",
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt (github flag enabled) 1`] = `
{
  "result": "
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-opt-spec-github-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
",
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "__tests__/__fixtures__/petstore-simple-weird-version.json is a valid OpenAPI API definition!

🚀 Let's get you set up with GitHub Actions! 🚀

",
}
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt (github flag enabled) 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-opt-spec-github-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-opt-spec-github-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v4

      - name: Run \`openapi validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi validate __tests__/__fixtures__/petstore-simple-weird-version.json
"
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt (including workingDirectory) 1`] = `
{
  "result": "
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-opt-spec-workdir-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
",
  "stderr": "- Validating the API definition located at petstore.json...
",
  "stdout": "petstore.json is a valid OpenAPI API definition!

🐙 Looks like you're running this command in a GitHub Repository! 🐙

🚀 With a few quick clicks, you can run this \`openapi validate\` command via GitHub Actions (https://github.com/features/actions)

✨ This means it will run automagically with every push to a branch of your choice!

",
}
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt (including workingDirectory) 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-opt-spec-github-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-opt-spec-github-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v4

      - name: Run \`openapi validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi validate petstore.json --workingDirectory=./__tests__/__fixtures__/relative-ref-oas
"
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt 1`] = `
{
  "result": "
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-opt-spec-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
",
  "stderr": "- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "__tests__/__fixtures__/petstore-simple-weird-version.json is a valid OpenAPI API definition!

🐙 Looks like you're running this command in a GitHub Repository! 🐙

🚀 With a few quick clicks, you can run this \`openapi validate\` command via GitHub Actions (https://github.com/features/actions)

✨ This means it will run automagically with every push to a branch of your choice!

",
}
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-opt-spec-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-opt-spec-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v4

      - name: Run \`openapi validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi validate __tests__/__fixtures__/petstore-simple-weird-version.json
"
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via prompts 1`] = `
{
  "result": "
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
",
  "stderr": "- Looking for API definitions...
- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "__tests__/__fixtures__/petstore-simple-weird-version.json is a valid OpenAPI API definition!

🐙 Looks like you're running this command in a GitHub Repository! 🐙

🚀 With a few quick clicks, you can run this \`openapi validate\` command via GitHub Actions (https://github.com/features/actions)

✨ This means it will run automagically with every push to a branch of your choice!

",
}
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via prompts 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v4

      - name: Run \`openapi validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi validate __tests__/__fixtures__/petstore-simple-weird-version.json
"
`;

exports[`rdme openapi validate > GHA onboarding E2E tests > should reject if user says no to creating GHA workflow 1`] = `
{
  "error": [Error: GitHub Actions workflow creation cancelled. If you ever change your mind, you can run this command again with the \`--github\` flag.],
  "stderr": "- Looking for API definitions...
- Validating the API definition located at __tests__/__fixtures__/petstore-simple-weird-version.json...
",
  "stdout": "__tests__/__fixtures__/petstore-simple-weird-version.json is a valid OpenAPI API definition!

🐙 Looks like you're running this command in a GitHub Repository! 🐙

🚀 With a few quick clicks, you can run this \`openapi validate\` command via GitHub Actions (https://github.com/features/actions)

✨ This means it will run automagically with every push to a branch of your choice!

",
}
`;

exports[`rdme openapi validate > error handling > should throw an error if an invalid API definition has many errors 1`] = `
{
  "error": [ValidationError: OpenAPI schema validation failed.

REQUIRED must have required property 'url'

   7 |   },
   8 |   "servers": [
>  9 |     {
     |     ^ url is missing here!
  10 |       "urll": "http://petstore.swagger.io/v2"
  11 |     }
  12 |   ],

ADDITIONAL PROPERTY must NOT have additional properties

  24 |     "/pet/findByStatus": {
  25 |       "get": {
> 26 |         "tagss": [
     |         ^^^^^^^ tagss is not expected to be here!
  27 |           "pet"
  28 |         ],
  29 |         "summary": "Finds Pets by status",],
  "stderr": "- Validating the API definition located at ./__tests__/__fixtures__/very-invalid-oas.json...
✖ Validating the API definition located at ./__tests__/__fixtures__/very-invalid-oas.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi validate > error handling > should throw an error if an invalid OpenAPI 3.0 definition is supplied 1`] = `
{
  "error": [ValidationError: API definition schema validation failed.

Missing $ref pointer "#/components/schemas/Error". Token "Error" does not exist.],
  "stderr": "- Validating the API definition located at ./__tests__/__fixtures__/invalid-oas.json...
✖ Validating the API definition located at ./__tests__/__fixtures__/invalid-oas.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi validate > error handling > should throw an error if an invalid OpenAPI 3.1 definition is supplied 1`] = `
{
  "error": [ValidationError: OpenAPI schema validation failed.

REQUIRED must have required property 'name'

  24 |   "components": {
  25 |     "securitySchemes": {
> 26 |       "tlsAuth": {
     |                  ^ name is missing here!
  27 |         "type": "mutualTLS"
  28 |       }
  29 |     }],
  "stderr": "- Validating the API definition located at ./__tests__/__fixtures__/invalid-oas-3.1.json...
✖ Validating the API definition located at ./__tests__/__fixtures__/invalid-oas-3.1.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi validate > error handling > should throw an error if an invalid Swagger definition is supplied 1`] = `
{
  "error": [ValidationError: Swagger schema validation failed.

ADDITIONAL PROPERTY must NOT have additional properties

  23 |         "description": "Find out more",
  24 |         "url": "http://swagger.io",
> 25 |         "this-shouldnt-be-here": true
     |         ^^^^^^^^^^^^^^^^^^^^^^^ this-shouldnt-be-here is not expected to be here!
  26 |       }
  27 |     }
  28 |   ],],
  "stderr": "- Validating the API definition located at ./__tests__/__fixtures__/invalid-swagger.json...
✖ Validating the API definition located at ./__tests__/__fixtures__/invalid-swagger.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi validate > error handling > should throw an error if invalid JSON is supplied 1`] = `
{
  "error": [SyntaxError: Unexpected end of JSON input],
  "stderr": "- Validating the API definition located at ./__tests__/__fixtures__/invalid-json/yikes.json...
✖ Validating the API definition located at ./__tests__/__fixtures__/invalid-json/yikes.json...
",
  "stdout": "",
}
`;

exports[`rdme openapi validate > should adhere to .gitignore in subdirectories 1`] = `
{
  "result": "nest/petstore.json is a valid OpenAPI API definition!",
  "stderr": "- Looking for API definitions...
- Validating the API definition located at nest/petstore.json...
",
  "stdout": "ℹ️  We found nest/petstore.json and are attempting to validate it.
",
}
`;

exports[`rdme openapi validate > should discover and upload an API definition if none is provided 1`] = `
{
  "result": "petstore.json is a valid OpenAPI API definition!",
  "stderr": "- Looking for API definitions...
- Validating the API definition located at petstore.json...
",
  "stdout": "ℹ️  We found petstore.json and are attempting to validate it.
",
}
`;

exports[`rdme openapi validate > should use specified working directory 1`] = `
{
  "result": "petstore.json is a valid OpenAPI API definition!",
  "stderr": "- Validating the API definition located at petstore.json...
",
  "stdout": "",
}
`;
